src/radiation/radiation_system.hpp

```diff
@@ -1169,9 +1188,80 @@ AMREX_GPU_HOST_DEVICE auto RadSystem<problem_t>::ComputeFluxInDiffusionLimit(con
+									amrex::GpuArray<double, nGroups_ + 1> const &rad_boundaries,
+									amrex::GpuArray<double, nGroups_> const &rad_boundary_ratios) -> double
+{
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	quokka::valarray<double, nGroups_> kappaEVec{};
+
+	const double num_density = rho / mean_molecular_mass_;
+	const double Lambda_compare = dustGasCoeff * num_density * num_density * std::sqrt(T_gas) * T_gas;
+
+	amrex::GpuArray<double, nGroups_> alpha_quant_minus_one{};
+	for (int g = 0; g < nGroups_; ++g) {
+		alpha_quant_minus_one[g] = -1.0;
+	}
+
+	// solve for dust temperature T_d using Newton iteration
+	double T_d = T_d_init;
+	const double lambda_rel_tol = 1.0e-8;
+	const int max_ite_td = 100;
+	int ite_td = 0;
+	for (; ite_td < max_ite_td; ++ite_td) {
+		const auto fourPiBoverC = ComputeThermalRadiation(T_d, rad_boundaries);
+
+		if constexpr (opacity_model_ == OpacityModel::single_group) {
+			kappaPVec[0] = ComputePlanckOpacity(rho, T_d);
+			kappaEVec[0] = ComputeEnergyMeanOpacity(rho, T_d);
+		} else {
+			const auto kappa_expo_and_lower_value = DefineOpacityExponentsAndLowerValues(rad_boundaries, rho, T_d);
+			if constexpr (opacity_model_ == OpacityModel::piecewise_constant_opacity) {
+				for (int g = 0; g < nGroups_; ++g) {
+					kappaPVec[g] = kappa_expo_and_lower_value[1][g];
+					kappaEVec[g] = kappa_expo_and_lower_value[1][g];
+				}
+			} else if constexpr (opacity_model_ == OpacityModel::PPL_opacity_fixed_slope_spectrum) {
+				kappaPVec = ComputeGroupMeanOpacity(kappa_expo_and_lower_value, rad_boundary_ratios, alpha_quant_minus_one);
+				kappaEVec = kappaPVec;
+			} else if constexpr (opacity_model_ == OpacityModel::PPL_opacity_full_spectrum) {
+				const auto alpha_B = ComputeRadQuantityExponents(fourPiBoverC, rad_boundaries);
+				const auto alpha_E = ComputeRadQuantityExponents(Erad, rad_boundaries);
+				kappaPVec = ComputeGroupMeanOpacity(kappa_expo_and_lower_value, rad_boundary_ratios, alpha_B);
+				kappaEVec = ComputeGroupMeanOpacity(kappa_expo_and_lower_value, rad_boundary_ratios, alpha_E);
+			}
+		}
+		AMREX_ASSERT(!kappaPVec.hasnan());
+		AMREX_ASSERT(!kappaEVec.hasnan());
+
+		const double LHS = c_light_ * rho * sum(kappaEVec * Erad - kappaPVec * fourPiBoverC) +
+				   dustGasCoeff * num_density * num_density * std::sqrt(T_gas) * (T_gas - T_d);
+
+		if (std::abs(LHS) < lambda_rel_tol * std::abs(Lambda_compare)) {
+			break;
+		}
+
+		const auto d_fourpib_over_c_d_t = ComputeThermalRadiationTempDerivative(T_d, rad_boundaries);
+		const double dLHS_dTd = -c_light_ * rho * sum(kappaPVec * d_fourpib_over_c_d_t) - dustGasCoeff * num_density * num_density * std::sqrt(T_gas);
+		const double delta_T_d = LHS / dLHS_dTd;
+		T_d -= delta_T_d;
+
+		if (ite_td > 0) {
+			if (std::abs(delta_T_d) < lambda_rel_tol * std::abs(T_d)) {
+				break;
+			}
+		}
+	}
+
+	AMREX_ASSERT_WITH_MESSAGE(ite_td < max_ite_td, "Newton iteration for dust temperature did not converge.");
```

Comment:
We can't handle errors this way on the GPU. This needs to be fixed.

Reply:
OK. I will solve it along with #643